File: typescript-driver/src/loader.ts
// Carga JS source descargado desde Postgres y devuelve instancia del driver.
// Usa node:vm con sandbox controlado: el driver NO tiene acceso a require, fs,
// ni globals de Node. Solo a console y APIs JS estándar.
//
// Diferencia con C# (AssemblyLoadContext): aquí no hay Unload real — V8 retiene
// los Script compilados hasta GC. Para hot-swap, descartar el handle y dejar
// que GC libere.

import * as crypto from 'node:crypto';
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as vm from 'node:vm';
import type { ComprobanteDriver } from './contract.js';

export class DriverHandle {
  constructor(
    public readonly version: string,
    public readonly hashSha256: string,
    public readonly instancia: ComprobanteDriver,
  ) {}

  static cargar(jsSource: string, version: string): DriverHandle {
    const hash = crypto.createHash('sha256').update(jsSource, 'utf-8').digest('hex');

    // Sandbox mínimo: console + JSON + Math + Date (puros). Sin require, fs, etc.
    const sandbox: Record<string, unknown> = {
      console,
      JSON,
      Math,
      Date,
      Number,
      String,
      Boolean,
      Array,
      Object,
      RegExp,
      Map,
      Set,
      Error,
      __exports: {} as Record<string, unknown>,
    };
    vm.createContext(sandbox);

    // El driver compilado debe asignar la clase a __exports.default o
    // __exports.<NombreClase>. tsc emite ESM/CommonJS — el compilar.ts lo
    // envuelve para que use __exports.
    const script = new vm.Script(jsSource, { filename: `driver-${version}.js` });
    script.runInContext(sandbox, { timeout: 5000 });

    const exports = sandbox.__exports as Record<string, unknown>;
    const candidato = (exports.default ?? Object.values(exports).find((v) => typeof v === 'function')) as
      | (new () => ComprobanteDriver)
      | undefined;
    if (!candidato) {
      throw new Error('No se encontró clase exportada que implemente ComprobanteDriver');
    }
    const inst = new candidato();
    if (typeof inst.preValidar !== 'function' || typeof inst.version !== 'string') {
      throw new Error('La clase cargada no implementa ComprobanteDriver (falta preValidar o version)');
    }
    return new DriverHandle(version, hash, inst);
  }
}

export class DriverCache {
  constructor(public readonly directorio: string) {
    fs.mkdirSync(directorio, { recursive: true });
  }
  private archivo(v: string): string { return path.join(this.directorio, `driver-${v}.js`); }
  tiene(v: string): boolean { return fs.existsSync(this.archivo(v)); }
  leer(v: string): string { return fs.readFileSync(this.archivo(v), 'utf-8'); }
  guardar(v: string, src: string): void { fs.writeFileSync(this.archivo(v), src, 'utf-8'); }
}

File: python-driver/src/chalona_driver/loader.py
"""Carga .py source descargado desde Postgres y devuelve instancia del driver.

Usa exec() en namespace dict aislado (ns). El driver descargado NO debe
intentar imports de chalona_driver.contract — el loader inyecta la clase
ComprobanteDriver directamente al ns para que las definiciones del driver
puedan heredar de ella.
"""
from __future__ import annotations

import hashlib
import inspect
import os
from pathlib import Path
from typing import Any

from .contract import ComprobanteDriver


class DriverHandle:
    def __init__(self, version: str, hash_sha256: str, instancia: ComprobanteDriver):
        self.version = version
        self.hash_sha256 = hash_sha256
        self.instancia = instancia

    @classmethod
    def cargar(cls, source: bytes, version: str) -> "DriverHandle":
        if isinstance(source, bytes):
            text = source.decode("utf-8")
        else:
            text = source
        h = hashlib.sha256(text.encode("utf-8")).hexdigest()

        # Namespace aislado con la base inyectada
        ns: dict[str, Any] = {
            "__name__": f"chalona_driver_dinamico_{version}",
            "__builtins__": __builtins__,
            "ComprobanteDriver": ComprobanteDriver,
        }
        compiled = compile(text, f"<driver-{version}>", "exec")
        exec(compiled, ns)

        candidato = None
        for name, obj in ns.items():
            if name.startswith("_") or name == "ComprobanteDriver":
                continue
            if inspect.isclass(obj) and issubclass(obj, ComprobanteDriver) and obj is not ComprobanteDriver:
                candidato = obj
                break
        if candidato is None:
            raise RuntimeError("No se encontró clase que herede de ComprobanteDriver")
        inst = candidato()
        return cls(version, h, inst)


class DriverCache:
    def __init__(self, directorio: str):
        self.directorio = Path(directorio)
        self.directorio.mkdir(parents=True, exist_ok=True)

    def _archivo(self, v: str) -> Path:
        return self.directorio / f"driver-{v}.py"

    def tiene(self, v: str) -> bool:
        return self._archivo(v).exists()

    def leer(self, v: str) -> bytes:
        return self._archivo(v).read_bytes()

    def guardar(self, v: str, src: bytes) -> None:
        self._archivo(v).write_bytes(src)
